feat: Add WASM host functions (Wasmi version)

mvadari opened the pull request:
## High Level Overview of Change

This PR adds all the Wasmi integration code and the host functions specified in XLS-102. It also adds tests for all of this.

_Note for reviewers: 100k lines of this PR are just WASM-compiled test fixtures, and another 5k lines are WASM test fixture source code. So while this PR is still quite large, it's not as large as it seems on the surface._

### Context of Change

This PR depends on #5999  and replaces #5791 

It is part of an effort to split up the Smart Escrow work into smaller, more managable-to-review PRs.

XLS-102: https://xls.xrpl.org/xls/XLS-0102-wasm-vm.html

Stack of PRs (from `develop` outwards):
* https://github.com/XRPLF/rippled/pull/5999
* This PR
* https://github.com/XRPLF/rippled/pull/5810
* https://github.com/XRPLF/rippled/pull/6329
* https://github.com/XRPLF/rippled/pull/5811

### Type of Change

<!--
Please check [x] relevant options, delete irrelevant ones.
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Performance (increase or change in throughput and/or latency)
- [ ] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

### API Impact

N/A

<!--
## Before / After
If relevant, use this section for an English description of the change at a technical level.
If this change affects an API, examples should be included here.

For performance-impacting changes, please provide these details:
1. Is this a new feature, bug fix, or improvement to existing functionality?
2. What behavior/functionality does the change impact?
3. In what processing can the impact be measured? Be as specific as possible - e.g. RPC client call, payment transaction that involves LOB, AMM, caching, DB operations, etc.
4. Does this change affect concurrent processing - e.g. does it involve acquiring locks, multi-threaded processing, or async processing?
-->

## Test Plan

Several tests are added to this PR.

<!--
## Future Tasks
For future tasks related to PR.
-->

mvadari committed:
add WAMR dependency

mvadari committed:
Add WAMR integration code

mvadari committed:
add host functions

mvadari committed:
add tests

mvadari committed:
CODEOWNERS

mvadari committed:
misc host function files

mvadari committed:
limits

mvadari committed:
bring out IOUAmount.h

mvadari committed:
add temBAD_WASM

mvadari committed:
update tests

mvadari committed:
add fixtures files

mvadari committed:
rearrange files

mvadari committed:
test: improve codecov in `HostFuncWrapper.cpp` (#5730)

mvadari committed:
chore: remove unneeded float stuff (#5729)

mvadari committed:
Merge branch 'develop' into ripple/wamr

mvadari committed:
Merge branch 'ripple/wamr' into ripple/wamr-host-functions

mvadari committed:
Merge branch 'develop' into ripple/wamr

mvadari committed:
Merge remote-tracking branch 'upstream/ripple/wamr' into wamr-host-functions

mvadari committed:
fix merge issues

oleks-rip committed:
Latests HF perf test (#5789)

mvadari committed:
Merge branch 'develop' into ripple/wamr

mvadari committed:
Merge branch 'ripple/wamr' into ripple/wamr-host-functions

oleks-rip committed:
Switch to own wamr fork (#5808)

mvadari committed:
Merge branch 'develop' into ripple/wamr

mvadari committed:
Merge branch 'ripple/wamr' into ripple/wamr-host-functions

mvadari committed:
Merge branch 'develop' into ripple/wamr

mvadari committed:
Merge branch 'ripple/wamr' into ripple/wamr-host-functions

mvadari committed:
switch from wasm32-unknown-unknown to wasm32v1-none (#5814)

mvadari committed:
Merge branch 'develop' into ripple/wamr

mvadari committed:
Merge branch 'ripple/wamr' into ripple/wamr-host-functions

mvadari committed:
Merge branch 'develop' into ripple/wamr

mvadari committed:
Merge branch 'ripple/wamr' into ripple/wamr-host-functions

mvadari committed:
fix: remove `get_ledger_account_hash` and `get_ledger_tx_hash` host functions (#5850)

* remove `get_ledger_account_hash` and `get_ledger_tx_hash`

* fix build+tests

mvadari committed:
Merge branch 'develop' into ripple/wamr

mvadari committed:
Merge branch 'ripple/wamr' into ripple/wamr-host-functions

mvadari committed:
Merge branch 'develop' into ripple/wamr

mvadari committed:
Merge branch 'ripple/wamr' into ripple/wamr-host-functions

oleks-rip committed:
Use xrplf conan repo for wamr (#5862)

mvadari committed:
Merge branch 'develop' into ripple/wamr

mvadari committed:
Merge branch 'ripple/wamr' into ripple/wamr-host-functions

mvadari committed:
Merge branch 'develop' into ripple/wamr

mvadari committed:
Merge branch 'ripple/wamr' into ripple/wamr-host-functions

mvadari committed:
Merge branch 'develop' into ripple/wamr

mvadari committed:
Merge branch 'ripple/wamr' into ripple/wamr-host-functions

mvadari committed:
return size from updateData

mvadari committed:
fix bug

mvadari committed:
fix tests

oleks-rip committed:
Chores: Sort package list (#5963)

mvadari committed:
Update cmake/RippledCore.cmake

Co-authored-by: Ayaz Salikhov <mathbunnyru@users.noreply.github.com>

mathbunnyru committed:
Merge branch 'develop' into ripple/wamr

oleks-rip committed:
Wamr and HF security review fixes (#5965)

mvadari committed:
allow 0-value trace amounts

mvadari committed:
update fixtures to use the latest version of stdlib

mvadari committed:
Merge branch 'develop' into wamr

mvadari committed:
Merge branch 'wamr' into wamr-host-functions

mvadari committed:
fix imports

mvadari committed:
run rename script

mvadari committed:
update

mvadari committed:
Merge branch 'ripple/wasmi' into wasmi-host-functions

mvadari committed:
fix build issue

mvadari committed:
Merge branch 'ripple/wasmi' into wasmi-host-functions

mvadari committed:
fix another build issue

mvadari committed:
Merge branch 'ripple/wasmi' into ripple/wasmi-host-functions

oleks-rip committed:
wasmi HF cost

pwang200 committed:
add memory limit and disable float and other advanced instructions

mvadari committed:
Merge remote-tracking branch 'upstream/ripple/wasmi' into wasmi-host-functions

mvadari committed:
fix parameter index desynchronization (#6148)

oleks-rip committed:
Fix potential memory leaks found by srlabs (#6145)

mvadari committed:
Merge branch 'ripple/wasmi' into ripple/wasmi-host-functions

mvadari committed:
merge fixes

mvadari committed:
fix more merge issues

mvadari committed:
fix more merge issues

pwang200 committed:
a batch of memory, table, and trap tests (#6100)

wasm memory, table, and trap unit tests

vvysokikh1 commented on src/xrpld/app/wasm/detail/HostFuncImpl.cpp:
```diff
@@ -0,0 +1,1305 @@
+        return data;
+    }
+};
+
```
I'm missing the point of having this class.

Most of it's functionality could be a part of base Number. The only part which isn't, `toBytes` function, could be a free function in this file, accepting number as an argument.

This also could mean that most of this could be a separate independent PR to develop, but I'm not forcing this in case you don't dant to do that

oleks-rip replied:
Conversion from Slice  shouldn't be part of the Number, uint64_t shouldn't be the part of the Number (as it can lead to silent conversion, and I don't want to check every usage of the numbers). `toBytes`  must be  private (static to this file) and its appearance in the class just because class is present here, and not to conflict with other possible  `toBytes`. 
It is good helper class only for  wasm "float" functions and will not interfere with the other modules

vvysokikh1 replied:
`Conversion from Slice shouldn't be part of the Number` - why?
`uint64_t shouldn't be the part of the Number (as it can lead to silent conversion, and I don't want to check every usage of the numbers)` - can be a named constructor (like `static Number fromUint64(...)`)
`toBytes` can be a static function in this file, I don't have any complaint to that.
`It is good helper class only for wasm "float" functions and will not interfere with the other modules` - I don't see any changes that you've made affecting other functionality of `Number`

oleks-rip replied:
> Conversion from Slice shouldn't be part of the Number

Because this slice is unique to "wasm float".  And someone can  misunderstand that and use it for core  Number serialization.

Overall there is no point in extending existing Number class with "wasm float" or exposing Number2 to the system. Especially with the latest changes to the mantissa handling

vvysokikh1 commented on src/libxrpl/basics/Number.cpp:
```diff
@@ -631,6 +631,48 @@ power(Number const& f, unsigned n)
+    auto const r = lnX / ln10;
+    return r;
+}
+
```
I see test covering these functions are in `TestHostFunctions.h`. Since you've added this to libxrpl, those tests should be here as well

vvysokikh1 replied:
Also could be a separate small PR to develop ideally, but not forcing this

oleks-rip replied:
[PR](https://github.com/XRPLF/rippled/pull/6268) with Number tests

mvadari committed:
infinite loop test (#6064)

mvadari committed:
Merge remote-tracking branch 'upstream/ripple/wasmi' into wasmi-host-functions

mvadari committed:
Merge branch 'ripple/wasmi' into wasmi-host-functions

pwang200 committed:
fix Uninitialized import entries lead to undefined behavior During WASM Instantiation

pwang200 committed:
section corruption unit tests

vvysokikh1 commented on clang_format.sh:
```diff
@@ -0,0 +1,19 @@
+  cd $basedir
+  $clangf -style=file -i "$file"
+  cd $dir
+done
```
I don't think this script should be in this PR and even not sure we want it in the repository at all

mvadari replied:
Agreed, removed

vvysokikh1 commented on include/xrpl/basics/Number.h:
```diff
@@ -339,6 +339,10 @@ abs(Number x) noexcept
 
+// logarithm with base 10
+Number
+lg(Number const& value);
```
While this is true in math world, in computer science `lg` typically means `log2`. This might be a bit confusing function name.

mvadari replied:
@oleks-rip

oleks-rip replied:
renamed to log10, [PR](https://github.com/XRPLF/rippled/pull/6268)

vvysokikh1 commented on src/libxrpl/basics/Number.cpp:
```diff
@@ -631,6 +631,48 @@ power(Number const& f, unsigned n)
+ln(Number const& x, unsigned iterations = 50)
+{
+    if (x <= 0)
+        throw std::runtime_error("Not positive value");
```
I don't like that we rely on `ln` to check for negative value. This should logically also be in `lg`. This is not a bug though, but potentially confusing

mvadari replied:
@oleks-rip

oleks-rip replied:
Added,[PR](https://github.com/XRPLF/rippled/pull/6268)

vvysokikh1 commented on src/libxrpl/basics/Number.cpp:
```diff
@@ -631,6 +631,48 @@ power(Number const& f, unsigned n)
+    Number const z = (x - 1) / (x + 1);
+    Number const zz = z * z;
+    Number denom = Number(1, -10);
+
```
Could you elaborate on this default value choice. I'm not an expert here and I find this value very obscure.

I asked AI and it suggested that proper tail approximation here could be `Number denom = Number(2 * iterations + 1);`. Take it with a grain of salt of course, but please at least provide some explanation on the choice in the comment

mvadari replied:
@oleks-rip

oleks-rip replied:
AI suggested it to be close to 0, small number.  It is not relevant anymore, as method changed

vvysokikh1 commented on src/libxrpl/basics/Number.cpp:
```diff
@@ -631,6 +631,48 @@ power(Number const& f, unsigned n)
+    return r;
+}
+
+Number
+lg(Number const& x)
+{
```
I'm pretty sure both ln and lg could be done constexpr. I'm not going to bother much since the only usage of it is from host function and that would not benefit from it

mvadari replied:
@oleks-rip

oleks-rip replied:
Not sure what do you mean, this function is using  at runtime, so it can't be constexpr.  If you mean "inline", then I just following the style -  `power` is twice as small and still not inlined

vvysokikh1 replied:
using constexpr does not restrict it from being used it the runtime. It allows to calculate these values in the compile time where it's possible to do so. In your case you would not benefit from this because of the way you use it, but it could be beneficial to someone using that in different conditions later. Not a high priority thing though, as initially pointed.

vvysokikh1 commented on include/xrpl/tx/wasm/HostFunc.h:
```diff
@@ -0,0 +1,517 @@
+#pragma once
```
These 2 functions `get_ledger_account_hash()` and `get_ledger_transaction_hash()` are still described in the specification, but there's no implementation. I don't remember if these 2 were already mentioned before

mvadari replied:
@mvadari

mvadari replied:
Looks like I updated the spec at some point and fixed this: https://xls.xrpl.org/xls/XLS-0102-wasm-vm.html

vvysokikh1 commented on src/xrpld/app/wasm/detail/HostFuncWrapper.cpp:
```diff
@@ -0,0 +1,2185 @@
+
+    return returnResult(
+        runtime, params, results, hf->checkKeylet(acc.value(), *seq), index);
+}
```
`hf->checkKeylet(acc.value(), *seq)` expects `uint32_t seq`, while seq is int32_t. I don't think it's possible to get negative sequence, but explicit cast, or change of seq type to match would be nice

mvadari replied:
@mvadari

mvadari replied:
Discussing the best solution to this

vvysokikh1 commented on src/xrpld/app/wasm/detail/HostFuncImpl.cpp:
```diff
@@ -0,0 +1,1305 @@
+    auto const& keylet = keylet::unchecked(objId);
+    if (cacheIdx < 0 || cacheIdx > MAX_CACHE)
+        return Unexpected(HostFunctionError::SLOT_OUT_RANGE);
+
```
I think it should be cacheIdx >= MAX_CACHE

mvadari replied:
@oleks-rip

oleks-rip replied:
It can be, cause in specification numbers [1;256] are specific slots, and 0 mean "use any free slot"
Later in code this value reduced by 1 to make correct [0; 256) index

vvysokikh1 commented on src/xrpld/app/wasm/detail/HostFuncImpl.cpp:
```diff
@@ -0,0 +1,1305 @@
+    Slice const& data,
+    bool asHex)
+{
+#ifdef DEBUG_OUTPUT
```
I think this function implementation should check what it prints (ie accept only letters/numbers/some symbol set). Right now this allows to print whatever wasm can insert (some binary data etc.). I would suggest we enforce this to be human-readable. I can't think of a way it could be abused though

mvadari replied:
It feels harder to enforce that it's human-readable

oleks-rip replied:
I believe this check should (if not already done) be in the LOG subsystem.

vvysokikh1 commented on src/xrpld/app/wasm/detail/HostFuncImpl.cpp:
```diff
@@ -0,0 +1,1305 @@
+        Number2 yy(y);
+        if (!yy)
+            return Unexpected(HostFunctionError::FLOAT_INPUT_MALFORMED);
+        Number2 res = xx / yy;
```
There's no check that yy can't be zero.

While the rippled does not crash, because the code in Number.cpp catches that, this is not the best solution.

mvadari replied:
@oleks-rip

oleks-rip replied:
This is exactly why we relay on Number   - to catch all overflows / divizion by zero / any other float errors

pwang200 committed:
fix start function loop

oleks-rip committed:
Add checks to allocate (#6185)

mvadari committed:
Merge branch 'ripple/wasmi' into ripple/wasmi-host-functions

mvadari committed:
Merge branch 'ripple/wasmi' into ripple/wasmi-host-functions

pwang200 committed:
disable Wasm features added in Wasmi 1.0, and fix unit test fuel cost due to Wasmi 1.0 fuel changes (#6173)

* disable 4 more wasm features

* unit tests for disabled Wasmi 1.0 features

* fix unit tests failed due to fuel changes

* rearrange wasm feature unit tests

* fix gas costs

* Update src/test/app/wasm_fixtures/wat/custom_page_sizes.wat

---------

Co-authored-by: Mayukha Vadari <mvadari@ripple.com>

oleks-rip committed:
Fix traces (#6127)

* Fix traces
* More tests for codecov
* Review fixes
* trace float test
* Fix return value for traces
* Remove SuiteJournalSink2
* Add explicit severity
* Move logs to ApplyView
* Add check for output strings
* Merging fix

oleks-rip committed:
Fix unaligned access (#6208)

mvadari committed:
move helper functions to separate file (#6178)

* move helper functions to separate file

* break it up into sections, split out float helpers

* split impls into multiple cpp files

* namespace detail

* fix build issue

* fix tests

* clean up

* put float helpers into wasm_float namespace

mvadari committed:
remove test file

mvadari committed:
Merge branch 'ripple/wasmi' into wasmi-host-functions

mvadari commented on .github/CODEOWNERS:
```diff
@@ -1,8 +1,2 @@
 # Allow anyone to review any change by default.
 *
-
```
NOTE: this change should be restored before merge

oleks-rip committed:
Fix merge issues

oleks-rip committed:
Fix usage of the  Number class

oleks-rip committed:
Don't throw from hostfunctions stack (#6221)

mvadari committed:
Add `Vector256` support to the locator (#6131)

* add Vector256 nesting/length support

* [WIP] add tests

* fix tests

* simplify with helper function

* oops typo

* remove static variable

* respond to comments

* STBaseOrUInt256->FieldValue

* oops

* add more tests for coverage

* respond to comments

mvadari committed:
improve trace hf code (#6190)

* adjust trace statements

* add helper function

* use lambda instead

* use same paradigm in TestHostFunctions

* oops

oleks-rip committed:
Ensure lifetime of imports (#6230)

mvadari committed:
Merge branch 'ripple/wasmi' into wasmi-host-functions

oleks-rip committed:
Check that max parameters length is multiple of sizeof(int32) (#6253)

mvadari committed:
minor refactor of Wasm_test (#6229)

mvadari committed:
update return calculation (#6250)

mvadari committed:
Merge branch 'ripple/wasmi' into ripple/wasmi-host-functions

oleks-rip committed:
Check wasm return type (#6240)

* Check wasm return type

* Add more tests

oleks-rip committed:
Switch to series expansion method for ln() (#6268)

* Switch to series expansion method for ln()
Add float lg() tests to Number tests;
* Rename lg -> log10
* Add check for 0 to log10()

mvadari committed:
clean up Wasm_test.cpp more (#6278)

oleks-rip committed:
Ensure request size less than int limit (#6239)

* Ensure request size less than int limit

* Move size check to wasmParams function

oleks-rip committed:
Fix for Big-Endian machines (#6245)

mvadari committed:
Merge branch 'ripple/wasmi' into ripple/wasmi-host-functions

oleks-rip committed:
Reject non-canonical binaries (#6277)

* Reject non-canonical binaries

* Review fixes

* Cleanup Number2 class

* Use enum instead of 0

mvadari committed:
Merge branch 'ripple/wasmi' into wasmi-host-functions

mvadari committed:
fix clang-format

mvadari committed:
Merge branch 'ripple/wasmi' into wasmi-host-functions

mvadari committed:
Merge branch 'ripple/wasmi' into wasmi-host-functions

mvadari committed:
use buffers for uint32 WASM params (#6291)

oleks-rip committed:
Fix negation of int64_t (#6296)

mvadari committed:
Merge branch 'ripple/wasmi' into wasmi-host-functions

mvadari committed:
add readme to src/xrpld/app/wasm (#6340)

* add readme to src/xrpl/app/wasm

* important block

* respond to copilot

mvadari committed:
Merge branch 'ripple/wasmi' into wasmi-host-functions

mvadari committed:
fix cspell issues in tests (#6348)

oleks-rip committed:
Add tests for wasm  functions with many parameters (#6343)

* Add functions with many parameters

* Add 10k locals function

* Module with  5k functions

* fix typo

Co-authored-by: Mayukha Vadari <mvadari@gmail.com>

---------

Co-authored-by: Mayukha Vadari <mvadari@gmail.com>

mvadari committed:
clean up some hf code (#6354)

* clean up some hf code

* fix comments

* fix ubsan

* Revert "fix ubsan"

mvadari committed:
Merge branch 'ripple/wasmi' into ripple/wasmi-host-functions

mvadari committed:
fix pre-commit

mvadari committed:
fix build issues

oleks-rip committed:
UT for wasm parameters (#6413)

oleks-rip committed:
Test invalid opcodes (#6392)

oleks-rip committed:
Disable reusing wasm module (#6364)

* Remove ability to re-use wasm module

* Check that HFS object is always new

* Fix clang format

* Remove perf tests

* temp build fix

* Fix merge

mvadari committed:
Merge branch 'ripple/wasmi' into ripple/wasmi-host-functions

mvadari committed:
Merge remote-tracking branch 'upstream/ripple/wasmi' into wasmi-host-functions

vvysokikh1 commented on src/libxrpl/tx/wasm/HostFuncWrapper.cpp:
```diff
@@ -0,0 +1,1939 @@
+
+    if (slice->size() == uint256::bytes)
+    {
+        if (auto ret = hf->isAmendmentEnabled(uint256::fromVoid(slice->data())); *ret == 1)
```
*ret without .has_value() is a dangerous code pattern. It could be UB potentially. While `isAmendmentEnabled` cannot return an error now, it's not guaranteed that this will not happen in the future. If the `hf` class is not set correctly (let's say you use base class for some reason), it will actually return an unexpected and instead of proper error handling we will have an UB

oleks-rip replied:
` if (auto ret ` check that it has value, cast to bool  is shorten form of  `ret.has_value() `

vvysokikh1 replied:
This is C++17's if-with-initializer syntax (https://en.cppreference.com/w/cpp/language/if.html):                                                                                                                                                                      
                                                                                                                                                                                                                   
  if (init-statement; condition) { ... }                                                                                                                                                                                                                                                                                                                                                                                    The statement splits into two distinct parts:                                                                                                                                              
  - Init statement: auto ret = hf->isAmendmentEnabled(...) — this just initializes ret. **It is not evaluated as a boolean**. It always succeeds regardless of what ret holds.                                       
  - Condition: *ret == 1 — this is the actual boolean check.                                                                                                                                                       
                                                                                                                                                                                                                   
  This is easy to confuse with the more familiar single-expression form:
                                                                                                                                                                                                                   
  if (auto ret = hf->isAmendmentEnabled(...))  // no semicolon — ret IS the condition

vvysokikh1 commented on src/libxrpl/tx/wasm/HostFuncWrapper.cpp:
```diff
@@ -0,0 +1,1939 @@
+    auto const* runtime = reinterpret_cast<InstanceWrapper const*>(hf->getRT());
+    int index = 0;
+
+    if (params->data[1].of.i32 + params->data[3].of.i32 > maxWasmParamLength)
```
I probably raised this before, but can this overflow? These two params are provided by the user right?

oleks-rip replied:
`maxWasmParamLength` is size_t , so not, but better to fix it to explicit form

vvysokikh1 replied:
The overflow happens in the addition itself, before any comparison and conversions.                                                                                                                                
                                                                                                                                                                                                                   
params->data[1].of.i32 and params->data[3].of.i32 are both int32_t. The addition int32_t + int32_t is performed in int32_t arithmetic

vvysokikh1 commented on src/libxrpl/tx/wasm/HostFuncImplFloat.cpp:
```diff
@@ -0,0 +1,550 @@
+            return;
+
+        Number::setround(static_cast<Number::rounding_mode>(mode));
+        Number::setMantissaScale(MantissaRange::mantissa_scale::small);
```
While I don't see the issue here, I see the comments in original implementation:
```
    /** Changes which mantissa scale is used for normalization.
     *
     * If you think you need to call this outside of unit tests, no you don't.
     */
    static void
    setMantissaScale(MantissaRange::mantissa_scale scale);
```

Could you bring this up for discussion?

mvadari replied:
@oleks-rip

oleks-rip replied:
we are using old form. I doubt  our 8 byte serialization format (used for exchange between wasm and c++) will fit extended mantissa

vvysokikh1 replied:
I would still like this to be resolved in a way where there's no blatant comment that it should not be used outside of unit-tests.

vvysokikh1 commented on src/xrpld/app/wasm/detail/HostFuncImplGetter.cpp:
```diff
@@ -0,0 +1,400 @@
+        if (STI_ARRAY == field->getSType())
+        {
+            auto const* arr = static_cast<STArray const*>(field);
+            if (sfieldCode >= arr->size())
```
is there any chance sfieldCode can be negative? This will cause an issue

mvadari replied:
Being fixed in #6376

oleks-rip replied:
fix already merged

vvysokikh1 commented on src/libxrpl/tx/wasm/HostFuncImplNFT.cpp:
```diff
@@ -0,0 +1,68 @@
+// =========================================================
+
+Expected<Bytes, HostFunctionError>
+WasmHostFunctionsImpl::getNFT(AccountID const& account, uint256 const& nftId) const
```
this function name is a bit misleading, should it be getNFTUri or something like that?

oleks-rip replied:
@mvadari

vvysokikh1 commented on src/libxrpl/tx/wasm/HostFuncWrapper.cpp:
```diff
@@ -0,0 +1,1939 @@
+
+    int64_t const gas = runtime->getGas();
+    WasmImportFunc const& impFunc = udata->second;
+    int64_t const x = gas >= impFunc.gas ? gas - impFunc.gas : 0;
+
+    if (runtime->setGas(x) < 0)
+    {
+        wasm_trap_t* trap = reinterpret_cast<wasm_trap_t*>(
+            WasmEngine::instance().newTrap("can't set gas"));  // LCOV_EXCL_LINE
+        return Unexpected(trap);                               // LCOV_EXCL_LINE
+    }
+
+    if (gas < impFunc.gas)
+    {
+        wasm_trap_t* trap =
+            reinterpret_cast<wasm_trap_t*>(WasmEngine::instance().newTrap("hf out of gas"));
+        return Unexpected(trap);
+    }
+
```
In case there's not enough gas, we first set gas to 0, and only then we check if there's enough gas to run the function. Is this intentional? Should we first check if there's enough and only then deduct?

mvadari replied:
@oleks-rip

oleks-rip replied:
I don't see any difference. In case there is not enough gas, we should set gas to 0 (function called, so we can charge) and should return an error.

vvysokikh1 commented on include/xrpl/tx/wasm/WasmVM.h:
```diff
@@ -0,0 +1,90 @@
+
+namespace xrpl {
+
+static std::string_view const W_ENV = "env";
+static std::string_view const W_HOST_LIB = "host_lib";
+static std::string_view const W_MEM = "memory";
+static std::string_view const W_STORE = "store";
+static std::string_view const W_LOAD = "load";
+static std::string_view const W_SIZE = "size";
+static std::string_view const W_ALLOC = "allocate";
+static std::string_view const W_DEALLOC = "deallocate";
+static std::string_view const W_PROC_EXIT = "proc_exit";
+
+static std::string_view const ESCROW_FUNCTION_NAME = "finish";
+
+uint32_t const MAX_PAGES = 128;  // 8MB = 64KB*128
+
```
This creates separate copies per translation unit. Use inline constexpr

mvadari replied:
@oleks-rip

oleks-rip replied:
As I remember consts are not copied but put altogether into "const data segment" of exe binaries, static should be removed

vvysokikh1 replied:
You are talking about the string data — the bytes "env", "host_lib", etc. Those literal bytes do live in the binary's read-only data segment, and linkers typically merge identical string literals. That part is correct.                                                                                                                                                                                        
                                                                                                                                                                                                                   But the string_view objects themselves — each a {pointer, size} pair — are a separate concern. With static at namespace scope in a header, each translation unit that includes the header gets its own internal-linkage copy of each string_view object.

Please use something like Protocol.h file as a reference about how to declare these constants

vvysokikh1 replied:
please use this doc as a guideline for why this is the correct way
https://abseil.io/tips/168

vvysokikh1 commented on src/libxrpl/tx/wasm/WasmVM.cpp:
```diff
@@ -0,0 +1,204 @@
+    auto const ret =
+        vm.run(wasmCode, hfs, funcName, params, createWasmImport(hfs), gasLimit, hfs.getJournal());
+
+    // std::cout << "runEscrowWasm, mod size: " << wasmCode.size()
+    //           << ", gasLimit: " << gasLimit << ", funcName: " << funcName;
```
nit: needs cleanup

mvadari replied:
@oleks-rip

vvysokikh1 commented on include/xrpl/tx/wasm/ParamsHelper.h:
```diff
@@ -0,0 +1,231 @@
+    {
+        std::int32_t i32;
+        std::int64_t i64 = 0;
+        float f32;
+        double f64;
```
These params are not used if I understand correctly, do you plan to use them later?

vvysokikh1 commented on src/libxrpl/tx/wasm/WasmiVM.cpp:
```diff
@@ -0,0 +1,923 @@
+
+    // Create and instantiate the module.
+    if (wasmCode.empty())
+        throw std::runtime_error("empty nodule");
```
nit: typo here

mvadari committed:
rename variable

mvadari committed:
Merge branch 'ripple/wasmi' into ripple/wasmi-host-functions

mvadari committed:
assorted fixes (#6376)

oleks-rip committed:
Fix clang tidy (#6463)

* Fix clang tidy

* Add exponent overflow test

a1q123456 committed:
Modularise wasm (#6441)

Signed-off-by: JCW <a1q123456@users.noreply.github.com>

mvadari committed:
Merge branch 'ripple/wasmi' into ripple/wasmi-host-functions

vvysokikh1 commented on include/xrpl/tx/wasm/WasmVM.h:
```diff
@@ -0,0 +1,90 @@
+    WasmEngine();
+
+    WasmEngine(WasmEngine const&) = delete;
+    WasmEngine(WasmEngine&&) = delete;
```
nit: it there any particular reason this cannot be moved?

oleks-rip replied:
we don't have such functionality, it is defensive check.

vvysokikh1 replied:
this class only has unique_ptr, which can be moved no problem. I was asking if there's a reason. Since there's no reason to that, what defensive check are you talking about?

vvysokikh1 commented on include/xrpl/tx/wasm/WasmiVM.h:
```diff
@@ -0,0 +1,321 @@
+    WasmValVec r;
+    bool f;  // failure flag
+
+    WasmiResult(unsigned N = 0) : r(N), f(false)
```
please make all 1 argument constructor `explicit`

oleks-rip replied:
It is ok if it converts implicitly

vvysokikh1 replied:
it's not and it's a source of many problems in c++. Please make it explicit

vvysokikh1 commented on include/xrpl/tx/wasm/WasmiVM.h:
```diff
@@ -0,0 +1,321 @@
+    {
+    }
+
+    ~WasmiResult() = default;
+    WasmiResult(WasmiResult&& o) = default;
+    WasmiResult&
+    operator=(WasmiResult&& o) = default;
```
please follow either rule of 0 or rule of 5

vvysokikh1 commented on include/xrpl/tx/wasm/WasmiVM.h:
```diff
@@ -0,0 +1,321 @@
+        beast::Journal j);
+
+public:
+    InstanceWrapper();
+
+    InstanceWrapper(InstanceWrapper&& o);
+
+    InstanceWrapper&
+    operator=(InstanceWrapper&& o);
+
+    InstanceWrapper(StorePtr& s, ModulePtr& m, WasmExternVec const& imports, beast::Journal j);
+
+    ~InstanceWrapper() = default;
+
```
also either rule of 0 or 5 please

vvysokikh1 replied:
also please reorder constructor to be on top

vvysokikh1 commented on include/xrpl/tx/wasm/WasmiVM.h:
```diff
@@ -0,0 +1,321 @@
+    getGas() const;
+
+    std::int64_t
+    setGas(std::int64_t) const;
```
this function logically should not be const?

oleks-rip replied:
As we are using C functions calls, they all shoudn't be const. But as it doesn't change C++ part let it be.

vvysokikh1 commented on include/xrpl/tx/wasm/HostFuncImpl.h:
```diff
@@ -0,0 +1,295 @@
+    checkSelf() const override
+    {
+        return !currentLedgerObj_ && !data_ &&
+            std::ranges::find_if(cache_, [](auto& p) { return !!p; }) == cache_.end();
```
this could a better readable option: 
```
  std::ranges::none_of(cache_, [](auto& p) { return static_cast<bool>(p); })
  // or even simpler:
  std::ranges::none_of(cache_, std::identity{})

vvysokikh1 commented on include/xrpl/tx/wasm/HostFuncImpl.h:
```diff
@@ -0,0 +1,295 @@
+
+    void const* rt_ = nullptr;
+
+    Expected<std::shared_ptr<SLE const>, HostFunctionError>
+    getCurrentLedgerObj() const
+    {
+        if (!currentLedgerObj_)
+            currentLedgerObj_ = ctx_.view().read(leKey_);
+        if (*currentLedgerObj_)
+            return *currentLedgerObj_;
+        return Unexpected(HostFunctionError::LEDGER_OBJ_NOT_FOUND);
+    }
+
```
I find this implementation quite dirty. Why does `currentLedgerObj_` be optional in the first place? It can be initialized in constructor and not be optional at all right?

oleks-rip replied:
It is about efficiency. It can be not used, so we only make lookup (which is very costly if not hit cache) when it actually needed.

vvysokikh1 replied:
I understand this is lazy call, but it does not mean this is a good way to do it? Escro will almost certainly access the current ledger object, so deferring the read saves nothing in practice. But accepting the premise, I'd rather go with separate bool flag (objLoaded_ or something)

vvysokikh1 commented on src/libxrpl/tx/wasm/HostFuncImplGetter.cpp:
```diff
@@ -0,0 +1,399 @@
+Bytes
+getIntBytes(STBase const* obj)
+{
+    static_assert(std::is_integral<T>::value, "Only integral types");
```
nit: could be `is_integral_v`

vvysokikh1 commented on src/libxrpl/tx/wasm/HostFuncImplGetter.cpp:
```diff
@@ -0,0 +1,399 @@
+{
+    static_assert(std::is_integral<T>::value, "Only integral types");
+
+    auto const& num(static_cast<STInteger<T> const*>(obj));
```
this creates unneccessary indirection.

I assume it should be either one of these:
```
auto const* num = static_cast<STInteger<T> const*>(obj);     
or                                                                                                                     
auto const& num = *static_cast<STInteger<T> const*>(obj);
```

vvysokikh1 commented on src/libxrpl/tx/wasm/HostFuncImplGetter.cpp:
```diff
@@ -0,0 +1,399 @@
+    static_assert(std::is_integral<T>::value, "Only integral types");
+
+    auto const& num(static_cast<STInteger<T> const*>(obj));
+    T const data = adjustWasmEndianess(num->value());
+    auto const* b = reinterpret_cast<uint8_t const*>(&data);
+    auto const* e = reinterpret_cast<uint8_t const*>(&data + 1);
+    return Bytes{b, e};
```
I think it would be easier to read as
```
auto const* b = reinterpret_cast<uint8_t const*>(&data); 
return Bytes{b, b + sizeof(T)};

vvysokikh1 commented on src/libxrpl/tx/wasm/HostFuncImplGetter.cpp:
```diff
@@ -0,0 +1,399 @@
+
+typedef std::variant<STBase const*, uint256 const*> FieldValue;
+
+namespace detail {
```
this should be an anonimous namespace most likely to not allow these functions pollute global symbol table or risk ODR

vvysokikh1 commented on src/libxrpl/tx/wasm/HostFuncImplGetter.cpp:
```diff
@@ -0,0 +1,399 @@
+getIntBytes(STBase const* obj)
+{
+    static_assert(std::is_integral<T>::value, "Only integral types");
+
```
the obj pointer is checked externally, but maybe worth putting an ASSERT here

vvysokikh1 commented on src/libxrpl/tx/wasm/HostFuncImplGetter.cpp:
```diff
@@ -0,0 +1,399 @@
+    if (cacheIdx >= MAX_CACHE)
+        return Unexpected(HostFunctionError::SLOTS_FULL);
+
+    cache_[cacheIdx] = ctx_.view().read(keylet);
+    if (!cache_[cacheIdx])
+        return Unexpected(HostFunctionError::LEDGER_OBJ_NOT_FOUND);
+    return cacheIdx + 1;  // return 1-based index
```
in an explicit id case when user provides an ID!=0, this can overwrite valid object with nullptr before returning an error. Is this intentional? Should we check if object is found before overwriting?

oleks-rip replied:
> this can overwrite valid object

It is just a cache, so ye, user can overwrite cached value on their behalf. 

> Should we check if object is found before overwriting?

I don't think someone expect untouched object after directly calling overwrite

vvysokikh1 replied:
but what's the point? If the object does not exist and you expect an error, why you'd want to evict existing object?

oleks-rip committed:
Add import / export sections test (#6497)

oleks-rip committed:
Clang-format fixtures.cpp (#6610)

mvadari committed:
Merge branch 'ripple/wasmi' into ripple/wasmi-host-functions

mvadari committed:
fix build

mvadari committed:
more build fixes

mvadari committed:
fix clang-tidy

mvadari committed:
more clang-tidy fixes

mvadari committed:
Merge branch 'ripple/wasmi' into ripple/wasmi-host-functions

mvadari committed:
clang-tidy fixes

oleks-rip committed:
Base divison of large fixtures (#6637)

mvadari committed:
test: Use proper length limits in codecov_tests (#6626)

mvadari committed:
Merge branch 'ripple/wasmi' of https://github.com/XRPLF/rippled into wasmi-host-functions

mvadari committed:
fix build

mvadari committed:
fix clang-tidy issues

oleks-rip committed:
fix Clang-tidy

mvadari committed:
fix clang-tidy issues

mvadari committed:
Merge branch 'ripple/wasmi' of https://github.com/XRPLF/rippled into wasmi-host-functions

mvadari committed:
Merge branch 'ripple/wasmi' of https://github.com/XRPLF/rippled into ripple/wasmi-host-functions

mvadari committed:
Merge branch 'ripple/wasmi' into ripple/wasmi-host-functions

mvadari committed:
fix build

oleks-rip committed:
New floats format, STAmount compatible (#6600)

mvadari committed:
Merge branch 'ripple/wasmi' of https://github.com/XRPLF/rippled into ripple/wasmi-host-functions

oleks-rip committed:
test: Calling wrap functions  from c++ side (#6699)

oleks-rip committed:
Review fixes (#6512)

oleks-rip commented on src/libxrpl/tx/wasm/HostFuncWrapper.cpp:
```diff
@@ -0,0 +1,1942 @@
+        if (auto ret = hf->isAmendmentEnabled(uint256::fromVoid(slice->data())); *ret == 1)
+        {
+            return returnResult(runtime, params, results, ret, index);
+        }
```
>  Expected (*ret) without checking ret has a value

fixed ce2586c039

>Also, for a 32-byte amendment id, the wrapper only returns when the result is 1; if the amendment is disabled (0) it falls through and treats the id bytes as a string, which is incorrect

planned behavior, support strings and binary representation

oleks-rip commented on include/xrpl/tx/wasm/README.md:
```diff
@@ -0,0 +1,189 @@
+> [!IMPORTANT]
+> New host functions MUST be amendment-gated in `WasmVM.cpp`.
+> Wrap the registration in an amendment check to ensure the function is only
+> available after the corresponding amendment is enabled on the network.
```
These ara original hostfunctions , not new

oleks-rip commented on src/libxrpl/tx/wasm/HostFuncWrapper.cpp:
```diff
@@ -0,0 +1,1942 @@
+
+    if (srcSize > maxWasmDataLength)
+        return HfErrorToInt(HostFunctionError::DATA_FIELD_TOO_LARGE);
+
```
Changed limit  to 1k, "returning error" is as planned

oleks-rip commented on src/test/app/TestHostFunctions.h:
```diff
@@ -0,0 +1,536 @@
+struct TestHostFunctionsSink : public TestHostFunctions
+{
+    test::StreamSink sink_;
+    void const* rt_ = nullptr;
```
fixed

pwang200 committed:
ai review nits fixes of host functions (#6963)

oleks-rip committed:
Refactoring float hostfunctions (#7053)

pwang200 committed:
bump get_nft host function cost from 1000 to 5000 (#7200)

pwang200 committed:
clarify XLS-0102 host function stability rule (#7146)

oleks-rip commented on src/test/app/wasm_fixtures/copyFixtures.py:
```diff
@@ -0,0 +1,125 @@
+    )
+    try:
+        subprocess.run(build_cmd, shell=True, check=True)
+        print(f"WASM file for {project_name} has been built and optimized.")
```
These are not part of the build system, just helper script to manual build. Fixtures always updated by author, not automatically

oleks-rip commented on src/test/app/wasm_fixtures/codecov_tests/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+panic = "abort"
+
+[dependencies]
+xrpl-std = { git = "https://github.com/ripple/xrpl-wasm-stdlib.git", package = "xrpl-wasm-stdlib", branch = "u32-buffer" }
```
These are not part of the build system, just helper script to manual build. Fixtures always updated by author, not automatically. There is no guarantee of reproducible build

oleks-rip commented on src/test/app/wasm_fixtures/all_keylets/Cargo.toml:
```diff
@@ -0,0 +1,21 @@
+panic = "abort"
+
+[dependencies]
+xrpl-std = { git = "https://github.com/ripple/xrpl-wasm-stdlib.git", package = "xrpl-wasm-stdlib", branch = "u32-buffer" }
```
These are not part of the build system, just helper script to manual build. Fixtures always updated by author, not automatically. There is no guarantee of reproducible build

oleks-rip commented on src/test/app/wasm_fixtures/all_host_functions/Cargo.toml:
```diff
@@ -0,0 +1,21 @@
+crate-type = ["cdylib"]
+
+[dependencies]
+xrpl-std = { git = "https://github.com/ripple/xrpl-wasm-stdlib.git", package = "xrpl-wasm-stdlib", branch = "u32-buffer" }
```
These are not part of the build system, just helper script to manual build. Fixtures always updated by author, not automatically. There is no guarantee of reproducible build

oleks-rip commented on src/test/app/TestHostFunctions.h:
```diff
@@ -0,0 +1,554 @@
+#include <test/app/wasm_fixtures/fixtures.h>
```
added in [PR#7000](https://github.com/XRPLF/rippled/pull/7000)

oleks-rip commented on src/test/app/wasm_fixtures/float_0/Cargo.toml:
```diff
@@ -0,0 +1,21 @@
+panic = "abort"
+
+[dependencies]
+xrpl-std = { git = "https://github.com/ripple/xrpl-wasm-stdlib.git", package = "xrpl-wasm-stdlib", rev = "1e5d096f46742ef7fcf1cb6f28a2526a72ed59d8" }
```
These are not part of the build system, just helper script to manual build. Fixtures always updated by author, not automatically. There is no guarantee of reproducible build

mvadari committed:
Merge branch 'ripple/wasmi' of https://github.com/XRPLF/rippled into ripple/wasmi-host-functions

mvadari committed:
fix clang-tidy issues

mvadari committed:
fix: Fix build issues post-clang-tidy changes (#7298)

oleks-rip committed:
Fix HF tests (#7365)

oleks-rip committed:
Cleanup and some refactoring (#7383)

oleks-rip committed:
HF one entry point (#7393)

Add one entry point for all HF for centralized exceptions handling, gas calculation and general checks.
Add exception handling for HF
Add FieldLocator object
Switch pointers to references for HF and runtime
Max size for parameters and sfData field is 1 kb now
Fix Allhf unittest, to provide correct locator

mvadari committed:
remove wasm engine tests

mvadari committed:
refactor: rename host functions (#7338)

Co-authored-by: xrplf-ai-reviewer[bot] <266832837+xrplf-ai-reviewer[bot]@users.noreply.github.com>

oleks-rip committed:
Memory transfer limit (#7000)

Count bytes copied across the boundaries (Wasm VM <-> Hostfunctions) and return error if limit reached (1 mb default)

pwang200 committed:
host function error path refactor (#7639)

TimothyBanks commented on include/xrpl/tx/wasm/WasmCommon.h:
```diff
@@ -0,0 +1,237 @@
+    virtual std::int64_t
+    setTransferLimit(std::int64_t transferLimit) = 0;
+};
+using RTOptRef = std::optional<std::reference_wrapper<WasmRuntimeWrapper>>;
```
This seems like a great place to use a type-erasure design on WasmRuntimeWrapper.  This would get us away from classical polymorphism (which is known to pollute type hierarchies and led to less flexible designs) and also minimize type definitions like this.

TimothyBanks replied:
You could also apply this to the host implementation types above.  This design would also allow you to only hook in the desired "overrides" instead of need to provide default implementations for all "virtual methods".

TimothyBanks commented on include/xrpl/tx/wasm/WasmCommon.h:
```diff
@@ -0,0 +1,237 @@
+inline void
+wasmParamsHlp(std::vector<WasmParam>& v, std::int32_t p, Types&&... args)
+{
+    v.push_back({.type = WasmTypes::WtI32, .of = {.i32 = p}});
```
Might be useful to just add some constructors to the WasmParam type and call `emplace_back` here.

TimothyBanks commented on include/xrpl/tx/wasm/WasmCommon.h:
```diff
@@ -0,0 +1,237 @@
+
+template <typename T, size_t Size = sizeof(T)>
+constexpr T
+adjustWasmEndianessHlp(T x)
```
What does this do over `std::byteswap` or boost variations?

TimothyBanks commented on include/xrpl/tx/wasm/WasmiVM.h:
```diff
@@ -0,0 +1,446 @@
+    T vec_;
+
+public:
+    WasmVec(size_t s = 0) : vec_ WASM_EMPTY_VEC
```
nit: can this be made explicit?

TimothyBanks commented on include/xrpl/tx/wasm/WasmiVM.h:
```diff
@@ -0,0 +1,446 @@
+
+namespace xrpl {
+
+template <class T, void (*Create)(T*, size_t), void (*Destroy)(T*)>
```
My opinion is that the last two template parameters would be better expressed as a policy design:
```
template <typename T>
struct lifecycle_policy {
  static void create(T* t, size_t size) { ... }
  static void destroy(T* t) { ... }
};

template <typename T, lifecycle_policy<T> policy>
class WasmVec
```

TimothyBanks commented on include/xrpl/tx/wasm/WasmiVM.h:
```diff
@@ -0,0 +1,446 @@
+    // WasmiEngine::call. std::nullopt means the call returned normally.
+    std::optional<TER> ter;
+
+    WasmiResult(unsigned n = 0) : r(n)
```
nit: can this be marked explicit?

TimothyBanks commented on include/xrpl/tx/wasm/WasmiVM.h:
```diff
@@ -0,0 +1,446 @@
+
+    InstanceWrapper(InstanceWrapper const&) = delete;
+
+    InstanceWrapper(InstanceWrapper&& o) : instance_(nullptr, &wasm_instance_delete)
```
Can `noexcept` be used on the move semantics in this type?

TimothyBanks commented on include/xrpl/tx/wasm/WasmiVM.h:
```diff
@@ -0,0 +1,446 @@
+    InstanceWrapper&
+    operator=(InstanceWrapper const&) = delete;
+
+    operator bool() const
```
It might be nice to pair this with a `has_value` function as well?

TimothyBanks commented on include/xrpl/tx/wasm/WasmiVM.h:
```diff
@@ -0,0 +1,446 @@
+
+    InstanceWrapper(InstanceWrapper&& o) : instance_(nullptr, &wasm_instance_delete)
+    {
+        *this = std::move(o);  // LCOV_EXCL_LINE
```
This seems a bit backwards idiomatically.  Normally assignment is implemented in terms of construction and swap.

TimothyBanks commented on include/xrpl/tx/wasm/WasmiVM.h:
```diff
@@ -0,0 +1,446 @@
+    {
+    }
+
+    ModuleWrapper(ModuleWrapper&& o) : module_(nullptr, &wasm_module_delete)
```
Can the move semantics be marked noexcept in this type?

TimothyBanks commented on include/xrpl/tx/wasm/WasmiVM.h:
```diff
@@ -0,0 +1,446 @@
+
+    ModuleWrapper(ModuleWrapper&& o) : module_(nullptr, &wasm_module_delete)
+    {
+        *this = std::move(o);
```
Similar question here about using move assignment within the move constructor

TimothyBanks commented on include/xrpl/tx/wasm/WasmiVM.h:
```diff
@@ -0,0 +1,446 @@
+
+public:
+    // LCOV_EXCL_START
+    ModuleWrapper() : module_(nullptr, &wasm_module_delete)
```
Can't `module_` be default member initialized and this constructor just be defaulted.  

Also, rule of 0 or 5 here.

TimothyBanks commented on include/xrpl/tx/wasm/WasmiVM.h:
```diff
@@ -0,0 +1,446 @@
+    std::mutex m_;  // 1 instance mutex
+
+public:
+    WasmiEngine() : engine_(init()), store_(nullptr, &wasm_store_delete)
```
Can these members be default member initialized and this constructor be defaulted?

Also, rule of 0 or 5 here.

TimothyBanks commented on include/xrpl/tx/wasm/WasmVM.h:
```diff
@@ -0,0 +1,89 @@
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+ImportVec
+createWasmImport(HostFunctions& hfs);
```
nit: I think for idiomatic cpp, it normally leans towards `makeWasmImport`.

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncImpl.cpp:
```diff
@@ -0,0 +1,52 @@
+Expected<Hash, HostFunctionError>
+WasmHostFunctionsImpl::computeSha512HalfHash(Slice const& data) const
+{
+    auto const hash = sha512Half(data);
```
nit:  Why the need for a temporary variable here?

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncImplFloat.cpp:
```diff
@@ -0,0 +1,594 @@
+        {
+            SerialIter it(data);
+            Number const x = STNumber(it, sfNumber).value();
+            *static_cast<Number*>(this) = x;
```
nit:  Similar comment here about calling the assignment operator within the constructor seems a bit anti pattern.  This seems like a double initialization.

This constructor seems like it should be more of a `makeWasmNumber` call that would return a `null` result if it fails.

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncImplFloat.cpp:
```diff
@@ -0,0 +1,594 @@
+            {
+                n = Number(static_cast<uint64_t>(mantissa), exponent, Number::Normalized{});
+            }
+            *static_cast<Number*>(this) = n;
```
nit:  same as the the one above about double init and this seems more like a `makeWasmNumber` method.

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncImplFloat.cpp:
```diff
@@ -0,0 +1,594 @@
+    operator=(WasmNumber const&) = default;
+
+    explicit
+    operator bool() const
```
This feels like the type is wanting to be some sort of an optional as well.  If the constructors above were make functions, we could just return an option instead of trying to bake those semantics into this type.

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncImplFloat.cpp:
```diff
@@ -0,0 +1,594 @@
+    {
+        std::string hex;
+        hex.reserve(data.size() * 2);
+        boost::algorithm::hex(data.begin(), data.end(), std::back_inserter(hex));
```
Isn't there already a method within xrpl that does this?

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncImplFloat.cpp:
```diff
@@ -0,0 +1,594 @@
+}
+
+Expected<Bytes, HostFunctionError>
+floatFromUintImpl(uint64_t x, int32_t mode)
```
This method is defined exactly the same as the one above it.  Can we template this and collapse it into one function?

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncImplFloat.cpp:
```diff
@@ -0,0 +1,594 @@
+}
+
+Expected<Bytes, HostFunctionError>
+floatFromSTAmountImpl(STAmount const& x, int32_t mode)
```
Same question as above.

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncImplFloat.cpp:
```diff
@@ -0,0 +1,594 @@
+}
+
+Expected<Bytes, HostFunctionError>
+floatFromSTNumberImpl(STNumber const& x, int32_t mode)
```
Same question as above

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncImplFloat.cpp:
```diff
@@ -0,0 +1,594 @@
+}
+
+Expected<int64_t, HostFunctionError>
+floatToIntImpl(Slice const& x, int32_t mode)
```
Same question as above

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncImplFloat.cpp:
```diff
@@ -0,0 +1,594 @@
+}
+
+Expected<Bytes, HostFunctionError>
+floatSubtractImpl(Slice const& x, Slice const& y, int32_t mode)
```
There appears to be a lot of methods with the exact same definition.  Can those be collapsed into one?

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncImplGetter.cpp:
```diff
@@ -0,0 +1,400 @@
+WasmHostFunctionsImpl::cacheLedgerObj(uint256 const& objId, int32_t cacheIdx)
+{
+    auto const& keylet = keylet::unchecked(objId);
+    if (cacheIdx < 0 || cacheIdx > maxCache)
```
Why not just use an unsigned integer?

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncImplGetter.cpp:
```diff
@@ -0,0 +1,400 @@
+        return Unexpected(sle.error());
+
+    auto const r = locateField(*sle.value(), locator);
+    if (!r)
```
nit:  I find it a bit inconsistent to have an `.has_value` check above and then a check using a conversion operator.  I feel like either both should be using `has_value` (probably `has_error` for this particular instance) or both should be using the conversion operator.

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncImplKeylet.cpp:
```diff
@@ -0,0 +1,222 @@
+    if (!account)
+        return Unexpected(HostFunctionError::InvalidAccount);
+    auto const keylet = keylet::account(account);
+    return Bytes{keylet.key.begin(), keylet.key.end()};
```
nit: There seems to be a mixture of uniform initialization and direct initialization.  Can we just standardize on uniform in this changeset?

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncImplKeylet.cpp:
```diff
@@ -0,0 +1,222 @@
+Expected<Bytes, HostFunctionError>
+WasmHostFunctionsImpl::checkKeylet(AccountID const& account, std::uint32_t seq) const
+{
+    if (!account)
```
There is definitely a pattern being defined by this definition and being reused within this file.  Could probably create a policy design pattern around it and encapsulate it that way.

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncImplTrace.cpp:
```diff
@@ -0,0 +1,71 @@
+Expected<int32_t, HostFunctionError>
+WasmHostFunctionsImpl::trace(std::string_view const& msg, Slice const& data, bool asHex) const
+{
+    if (!asHex)
```
It seems like the `asHex` flag is making this function do too much.  If the caller wants to see hex data, why can't they encode it that way in the `data` parameter?  Then any encoding could be used like base64 as well.

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncWrapper.cpp:
```diff
@@ -0,0 +1,1869 @@
+    int64_t const x = transLimit >= delta ? transLimit - delta : 0;
+
+    if (rt.setTransferLimit(x) < 0)
+        Throw<std::runtime_error>(std::string(hfErrInternal));  // LCOV_EXCL_LINE
```
Is there a reason that these throws need to create a `std::string`?  I would assume these string_view variables would outlive the scope of the throw?

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncWrapper.cpp:
```diff
@@ -0,0 +1,1869 @@
+getDataInt32(WasmRuntimeWrapper const&, wasm_val_vec_t const* params, int32_t& i)
+{
+    auto const result = params->data[i].of.i32;
+    i++;
```
nit:  This is entirely subjective, but I feel like the default should be pre increment unless there is a reason for post increment.  I get because this is an integral type it will be identical code, but it is one less thing to worry about.

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncWrapper.cpp:
```diff
@@ -0,0 +1,1869 @@
+    }
+    else
+    {
+        x = *reinterpret_cast<T const*>(r->data());
```
This is a strict aliasing violation and the alignment guard above only protects against alignment UB.  There are still possible aliasing and lifetime violations that can occur here.

The `memcpy` is the correct path for both branches.

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncWrapper.cpp:
```diff
@@ -0,0 +1,1869 @@
+static Expected<Currency, HostFunctionError>
+getDataCurrency(WasmRuntimeWrapper& runtime, wasm_val_vec_t const* params, int32_t& i)
+{
+    auto const slice = getDataSlice(runtime, params, i);
```
This definition is almost exactly the same as the previous two function definitions.  Can this be encapsulated in a policy design pattern to minimize code duplication?

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncWrapper.cpp:
```diff
@@ -0,0 +1,1869 @@
+        return locator;
+    }
+
+    int32_t const* locPtr = reinterpret_cast<int32_t const*>(slice->data());
```
This is a borderline strict aliasing violation as well.  `memcpy` is the right choice for both branches.

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncWrapper.cpp:
```diff
@@ -0,0 +1,1869 @@
+        if (index < 0 || index + 1 >= params->size)
+            Throw<std::runtime_error>(std::string(hfErrInternal));  // LCOV_EXCL_LINE
+
+        auto const resultValue = adjustWasmEndianess(res.value());
```
There seems to be code duplication across these branches.  Can we encapsulate this using a policy design or collapse the exact duplications into one branch?

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncWrapper.cpp:
```diff
@@ -0,0 +1,1869 @@
+wasm_trap_t*
+getCurrentLedgerObjField_wrap(WASM_SECONDARY_CB_PARAMS_LIST)
+{
+    int index = 0;
```
There is also a common pattern across these functions.  A templated policy design implementation would reduce that.

TimothyBanks commented on src/libxrpl/tx/wasm/HostFuncWrapper.cpp:
```diff
@@ -0,0 +1,1869 @@
+}
+
+// LCOV_EXCL_START
+namespace test {
```
Can these tests be moved into a test only source file?

TimothyBanks commented on src/libxrpl/tx/wasm/WasmiVM.cpp:
```diff
@@ -0,0 +1,957 @@
+    {
+        // wasm_trap_message NUL-terminates, so drop the trailing NUL.
+        out.assign(msg.data, msg.size - 1);
+        wasm_byte_vec_delete(&msg);
```
nit:  Wrap the `wasm_byte_vec_t` within an RAII type.

TimothyBanks commented on src/libxrpl/tx/wasm/WasmiVM.cpp:
```diff
@@ -0,0 +1,957 @@
+        }
+
+        if (errorMessage.size != 0u)
+            wasm_byte_vec_delete(&errorMessage);
```
nit:  Wrap the `wasm_byte_vec_t` within an RAII type.

TimothyBanks commented on src/libxrpl/tx/wasm/WasmiVM.cpp:
```diff
@@ -0,0 +1,957 @@
+        return *this;  // LCOV_EXCL_LINE
+
+    store_ = o.store_;
+    o.store_ = nullptr;
```
```suggestion
    store_ = std::exchange(o.store_, nullptr);
```

TimothyBanks commented on src/libxrpl/tx/wasm/WasmiVM.cpp:
```diff
@@ -0,0 +1,957 @@
+    o.store_ = nullptr;
+    exports_ = std::move(o.exports_);
+    memIdx_ = o.memIdx_;
+    o.memIdx_ = -1;
```
```suggestion
    memIdx_ = std::exchange(o.memIdx_, -1);
```

TimothyBanks commented on src/libxrpl/tx/wasm/WasmiVM.cpp:
```diff
@@ -0,0 +1,957 @@
+        return -1;  // LCOV_EXCL_LINE
+    std::uint64_t gas = 0;
+    wasm_store_get_fuel(store_, &gas);
+    return static_cast<std::int64_t>(gas);
```
Does this mean we can have negative gas at some point?

TimothyBanks commented on src/libxrpl/tx/wasm/WasmiVM.cpp:
```diff
@@ -0,0 +1,957 @@
+    if (store_ == nullptr)
+        return -1;  // LCOV_EXCL_LINE
+
+    if (gas < 0)
```
Why not just use an unsigned type?

TimothyBanks commented on src/libxrpl/tx/wasm/WasmiVM.cpp:
```diff
@@ -0,0 +1,957 @@
+{
+    if (store_ == nullptr)
+        return -1;  // LCOV_EXCL_LINE
+    if (x < 0)
```
Why not just use an unsigned type?

TimothyBanks commented on src/libxrpl/tx/wasm/WasmiVM.cpp:
```diff
@@ -0,0 +1,957 @@
+    bool instantiate,
+    ImportVec const& imports,
+    beast::Journal j)
+    : module_(init(s, wasmBin, j)), j_(j)
```
super nit:  Why not use uniform initialization?

TimothyBanks commented on src/libxrpl/tx/wasm/WasmiVM.cpp:
```diff
@@ -0,0 +1,957 @@
+        std::unique_ptr<wasm_functype_t, decltype(&wasm_functype_delete)> const ftype(
+            wasm_functype_new(params.get(), results.get()), &wasm_functype_delete);
+
+        params.release();
```
Is it possible for the construction of `ftype` to throw and yet it has taken ownership over these handles which would be a double delete.

TimothyBanks commented on src/libxrpl/tx/wasm/WasmiVM.cpp:
```diff
@@ -0,0 +1,957 @@
+    store_.reset();  // to free the memory before creating new store
+    store_ = {wasm_store_new_with_memory_max_pages(engine_.get(), maxPages), &wasm_store_delete};
+
+    if (gas < 0)
```
Why not just use an unsigned type?

TimothyBanks commented on src/libxrpl/tx/wasm/WasmiVM.cpp:
```diff
@@ -0,0 +1,957 @@
+static inline uint64_t
+usecs()
+{
+    uint64_t x = std::chrono::duration_cast<std::chrono::microseconds>(
```
Should something like this be tied to block time?  Or block number?

TimothyBanks commented on src/libxrpl/tx/wasm/WasmiVM.cpp:
```diff
@@ -0,0 +1,957 @@
+    ImportVec const& imports,
+    beast::Journal j)
+{
+    if (gas <= 0)
```
The last time I will ask about this, but why not just make this an unsigned integer?

TimothyBanks commented on src/libxrpl/tx/wasm/WasmiVM.cpp:
```diff
@@ -0,0 +1,957 @@
+    beast::Journal j)
+{
+    // currently only 1 module support, possible parallel UT run
+    std::scoped_lock const lg(m_);
```
This is a pretty large lock, could some things be moved outside of the lock (like the pre conditions on the input parameters).

TimothyBanks commented on src/libxrpl/tx/wasm/WasmiVM.cpp:
```diff
@@ -0,0 +1,957 @@
+    beast::Journal j)
+{
+    // currently only 1 module support, possible parallel UT run
+    std::scoped_lock const lg(m_);
```
Similar question on this lock.  Can the scope of this lock be decreased.  For instance moving the pre condition checks on the inputs outside of it.

TimothyBanks commented on src/libxrpl/tx/wasm/WasmVM.cpp:
```diff
@@ -0,0 +1,219 @@
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+WasmEngine::WasmEngine() : impl_(std::make_unique<WasmiEngine>())
```
I feel that a type erasure pattern would be much more flexible than the pimpl pattern.

mvadari committed:
Merge branch 'ripple/wasmi' of https://github.com/XRPLF/rippled into ripple/wasmi-host-functions

mvadari committed:
fix build issues

mvadari committed:
fix clang-tidy issues

mvadari committed:
feat: Change function name from `finish` to `escrow_finish` (#7840)

mvadari committed:
Merge branch 'ripple/wasmi-host-functions' into ripple/se/hf-merge-july

mvadari committed:
refactor: Remove WasmNumber class (#7833)

mvadari committed:
Merge remote-tracking branch 'upstream/ripple/wasmi' into ripple/se/hf-merge-july

mvadari committed:
Merge remote-tracking branch 'upstream/ripple/wasmi-host-functions' into ripple/se/hf-merge-july

mvadari committed:
fix build

mvadari committed:
fix test

kuznetsss committed:
Merge develop

kuznetsss committed:
chore: Fix clang-tidy issues

mvadari committed:
Merge pull request #7851 from kuznetsss/Update_hf-merge-july

chore: Update hf merge july

mvadari committed:
Merge pull request #7834 from XRPLF/ripple/se/hf-merge-july

chore: Merge in from `ripple/wasmi`

mvadari committed:
Merge branch 'ripple/wasmi' into ripple/wasmi-host-functions

kuznetsss committed:
Merge branch 'ripple/wasmi' into ripple/wasmi-host-functions
